src/orion/client/runner.py

```diff
@@ -0,0 +1,334 @@
+            new_trials = self._suggest_trials(n_trial)
+            log.debug(f"Sampled {len(new_trials)} new configs")
+            return new_trials
+
+        return []
+
+    def scatter(self, new_trials):
+        """Schedule new trials to be computed"""
+        new_futures = []
+        for trial in new_trials:
+            future = self.client.executor.submit(
+                _optimize, trial, self.fct, self.trial_arg, **self.kwargs
+            )
+            self.pending_trials[future] = trial
+            new_futures.append(future)
+
+        self.futures.extend(new_futures)
+        log.debug("Scheduled new trials")
+
+    def gather(self):
+        """Gather the results from each worker asynchronously"""
+        results = []
+        try:
+            results = self.client.executor.async_get(
+                self.futures, timeout=self.gather_timeout
+            )
+        except KeyboardInterrupt:
+            self._release_all()
+            raise
+
+        to_be_raised = None
+        log.debug(f"Gathered new results {len(results)}")
+
+        # register the results
+        for result in results:
+            trial = self.pending_trials.pop(result.future)
+
+            if isinstance(result, AsyncResult):
+                # NB: observe release the trial already
+                self.client.observe(trial, result.value)
+                self.trials += 1
+
+            if isinstance(result, AsyncException):
+                if (
+                    isinstance(result.exception, ExecutionError)
+                    and result.exception.return_code == self.interrupt_signal_code
+                ):
+                    to_be_raised = KeyboardInterrupt()
+                    self.client.release(trial, status="interrupted")
+                    continue
+
+                if isinstance(result.exception, InvalidResult):
+                    # stop the optimization process if we received `InvalidResult`
+                    # as all the experiments are assumed to be returning those
+                    to_be_raised = result.exception
+
+                # Regular exception, might be caused by the choosen hyperparameters
+                # themselves rather than the code in particular (like Out of Memory error
+                # for big batch sizes)
+                exception = result.exception
+                self.worker_broken_trials += 1
+                self.client.release(trial, status="broken")
+
+                if self.on_error is None or self.on_error(
+                    self, trial, exception, self.worker_broken_trials
+                ):
+                    log.error(result.traceback)
+
+                else:
+                    log.error(str(exception))
+                    log.debug(result.traceback)
+
+                # if we receive too many broken trials, it might indicate the user script
+                # is broken, stop the experiment and let the user investigate
+                if self.is_broken:
```

Comment:
This [comment](https://github.com/Epistimio/orion/pull/684#discussion_r762084215) still applies here.

Reply:
I cannot see the linked comment, can you copy paste it here ?

Reply:
Oops, me neither. I managed to find it back:
> We should wait and loop over all results before raising this error otherwise we may loose information of completed trials.

Reply:
updated

Reply:
Alright. There should also be a test to verify that calling gather one 1 broken and 1 completed trial does register the completed trial properly even though the one before in the list was broken with `to_be_raised is not None`.

Reply:
added test_multi_results_with_failure

Reply:
Very good, thanks!